improver/calibration/rainforest_calibration.py

```diff
@@ -0,0 +1,170 @@
+        an associated path has been supplied for all thresholds, otherwise lightgbm the
+        Boosters are returned.
+
+        Returns:
+            - numpy array containing the error threshold values.
+            - list of tree-model objects to be used in calibration process.
+        """
+        import importlib
+
+        from lightgbm import Booster
+
+        if importlib.util.find_spec("treelite") is not None:
+            from treelite_runtime import Predictor
+
+            self.treelite_enabled = True
+        else:
+            warnings.warn(
+                "Module treelite_runtime unavailable. Defaulting to using lightgbm Boosters."
+            )
+            self.treelite_enabled = False
+
+        error_thresholds = list(model_config_dict.keys())
+
+        lightgbm_model_filenames = [
+            model_config_dict[threshold].get("lightgbm_model")
+            for threshold in error_thresholds
+        ]
+        treelite_model_filenames = [
+            model_config_dict[threshold].get("treelite_model")
+            for threshold in error_thresholds
+        ]
+        if (None not in treelite_model_filenames) and self.treelite_enabled:
+            self.tree_models = [
+                Predictor(libpath=file, verbose=False, nthread=nthreads)
+                for file in treelite_model_filenames
+            ]
+        else:
+            self.tree_models = [
+                Booster(model_file=file).reset_parameter({"num_threads": nthreads})
+                for file in lightgbm_model_filenames
+            ]
+
+        self.error_thresholds = np.array(error_thresholds, dtype=np.float32)
+
+    def process(
+        self,
+        forecast_cube: Cube,
+        feature_cubes: CubeList,
+        error_percentiles_count: int = DEFAULT_ERROR_PERCENTILES_COUNT,
+        output_realizations_count: int = DEFAULT_OUTPUT_REALIZATIONS_COUNT,
+    ) -> Cube:
+        """Apply rainforests calibration to forecast cube.
+
+        The calibration is done in a situation dependent fashion using a series of
+        decision-tree models to construct representative error distributions which are
+        then used to map each input ensemble member onto a series of realisable values.
+
+        These error distributions are formed in a two-step process:
+
+        1. Evalute error CDF defined over the specified error_thresholds. Each exceedence
+        probability is evaluated using the corresponding decision-tree model.
+
+        2. Interpolate the CDF to extract a series of percentiles for the error distribution.
+        The error percentiles are then applied to each associated ensemble realization to
+        produce a series of realisable values; collectively these series form a calibrated
+        super-ensemble, which is then sub-sampled to provide the calibrated forecast.
+
+        Args:
+            forecast_cube:
+                Cube containing the forecast to be calibrated.
+            feature_cubes:
```

Comment:
I assume feature cubes are expected to have the same dimensions as the forecast cube, but represent different parameters.

Reply:
Yeah, though there is the possibility here that some features (derived fields) can have no realization dimension. In processing, these cubes are broadcast to match the cubes that do contain realization dimension.

I'll update the docstring to describe this behaviour in more detail.

Reply:
I've updated the docstring to be more explicit; hopefully this is clearer now, but would appreciate any suggestions.

Reply:
I think this is pretty clear now.